### ISSUE TITLE
Favicon via API

### ISSUE THREAD

**Issue by dbuechel** [MicrosoftEdge/WebView2Feedback#615]
We'd like to propose an extension of the API to easily retrieve the current favicon of a web page (something like https://cefsharp.github.io/api/85.3.x/html/M_CefSharp_IDisplayHandler_OnFaviconUrlChange.htm). Ideally, this API would also handle the case where a web page dynamically changes its favicon over time.

[AB#30216493](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/30216493)

**Comment by champnic** [comment]
I've added this request to our backlog. Thanks!

**Comment by darbid** [comment]
Just landed here to find this one on the feature request list.  It would be nice if this was implemented.

**Comment by jasonstephen15** [comment]
Hi @dbuechel / @darbid, we are investigating this feature request now. Do both of you mind expanding a bit on the scenario of how you would use such API?

**Comment by darbid** [comment]
My use case is I save a session history of urls visited. In my list I show the fav icon of the URL. I get the icon when the user visits the url and not when the user later shows the list so I avoid any login or authentication issues.

**Comment by dbuechel** [comment]
Our use case is that we want to render the favicon as window icon or aside the title of a browser tab, as most browsers nowadays do.

**Comment by ukandrewc** [comment]
There is thisL https://www.google.com/s2/favicons?domain=

**Comment by darbid** [comment]
@ukandrewc that does not work for some sites and as far as I can tell there is no way to detect it did not work as it returns a default icon.. it's my last choice.

**Comment by ukandrewc** [comment]
Ah yes, I clicked the link, but didn't notice it wasn't GitHub icon ;-)

This gets the URL or default:
```
const l=document.evaluate("//link[contains(@rel,'icon')]", document).iterateNext()
if(l==null)
  document.location.protocol+"//"+document.location.host+"/favicon.ico"
else
  x.href
```

**Comment by darbid** [comment]
Actually current best work around is to follow some code like Microsoft shows in their own examples see the HandleTabNavCompleted line 562 [here](https://github.com/MicrosoftEdge/WebView2Browser/blob/master/BrowserWindow.cpp).  The only issue with it is that it looks for rel=icon, but some websites now have a svg resource attached to this element which is a little difficult at least with WPF to deal with.  For example look at Github. 

```
<link rel="alternate icon" class="js-site-favicon" type="image/png" href="https://github.githubassets.com/favicons/favicon.png">

<link rel="icon" class="js-site-favicon" type="image/svg+xml" href="https://github.githubassets.com/favicons/favicon.svg">
```

You need to extend the javascript MS wrote to check for the image type and/or check for rel=alternate icon in the case of Github.

**Comment by ukandrewc** [comment]
Why the 'Actually' attitude ;-(
They can also be emdedded base64

**Comment by darbid** [comment]
> @darbid Why the 'Actually' attitude ;-(
> They can also be emdedded base64

I edited my text.

**Comment by ukandrewc** [comment]
@darbid  Thank you, appreciated ;-)

**Comment by Tochibee** [comment]
@dbuechel  Hey, I've been tasked to look into the implementation of this feature. I do have a question though, could you elaborate on what you mean by  "this API would also handle the case where a web page dynamically changes its favicon over time".  Are you simply referring to if the current webpage were to change its favicon while the page is active, that you would be able to get the url of the new icon, or something else?

**Comment by dbuechel** [comment]
@Tochibee Excellent, thanks a lot. Yes indeed, there are web pages which update their favicon to indicate a status change or alike (e.g. a build server once the build is completed or a chat client when a new message has arrived). Ideally, this API would handle that case by firing an event. So for us it would be sufficient if the API were an event which is fired as follows:

- never (if the page has no favicon)
- once during page load (if it's a static favicon)
- multiple times (when the favicon changes over time).

**Comment by Tochibee** [comment]
Ah, that makes sense. thanks for the clarification.

**Comment by Tochibee** [comment]
Hey @darbid  and @dbuechel  and anyone else. Is having the favicon URL all that you need, or will it also be necessary/beneficial to return the image itself in addition to the URL?

**Comment by dbuechel** [comment]
@Tochibee It certainly would be very beneficial to have the image itself as well, at least for us, then we wouldn't need to load it manually ourselves.

**Comment by darbid** [comment]
@Tochibee the url is a great start but for some urls you would need to add cookies etc.
, so I would very much like the image. However as I wrote above if their is no fav icon or your code fails please don't return a default icon.

**Comment by leonidukg** [comment]
Need to return exactly the finished picture of the site. The connection can work through a proxy and you can't request the picture directly. And you should also store the favicon in the cache, so you don't have to ask the site about it every time.

**Comment by champnic** [comment]
The CoreWebView2.FaviconChanged event is now available for use in SDKs 1.0.1293.44+. Thanks!
